README.md
# Podinfo

## Table of contents

- [Overview][1]

- [Code Statistics][11]

- [Usage and Demo][10]

  - [Docker Compose Environment][13]
  - [Build from source][14]

- [Build Systems][2]

  - [Build Systems : Overview][3]
  - [Build Systems : mage][4]
  - [Build Systems : just][5]
  - [Build Systems : docker buildx][6]

- [Roadmap][12]

## Overview

> TODO

```console
❯ podinfo server -h
Usage: podinfo server [options]

Starts podinfo server.

Telemetry Options:

  -metrics-prefix=<value>
     flag used to set default metrics prefix.
     This can also be specified via the 'PODINFO_METRICS_PREFIX' env
     variable

  -prometheus-retention-time=<value>
     flag used to set prometheus retention time.
     This can also be specified via the
     'PODINFO_PROMETHEUS_RETENTION_TIME' env variable

  -statsd-addr=<value>
     flag used to set statsd address
     This can also be specified via the 'STATSD_ADDR' env variable

  -statsite-addr=<value>
     flag used to set statsite address.
     This can also be specified via the 'STATSITE_ADDR' env variable

Server Options:

  -api-addr=<value>
     flag used to set the address podinfo is listening on.
     This can also be specified via the 'PODINFO_API_ADDR' env variable.

  -dev
     flag used to start the server in development mode
     This can also be specified via the 'PODINFO_DEVEL' env variable
     (true|false)

  -disable-cache
     flag used to disable redis cache and related endpoints
     This can also be specified via the 'PODINFO_DISABLE_CACHE' env
     variable (true|false)

  -log-level=<value>
     flag used to set stdlogger level.
     This can also be specified via the 'PODINFO_LOG_LEVEL' env
     variable.

  -node-name=<value>
     flag used to set podinfo node name.
     This can also be specified via the 'PODINFO_NODE_NAME' env
     variable.

Redis Options:

  -redis-addr=<value>
     flag used to set  address
     it is in host:port address format.
     This can also be specified via the 'PODINFO_REDIS_ADDR' env
     variable

  -redis-client-name=<CLIENT SETNAME ClientName>
     flag used to set  client name.
     ClientName will execute the `CLIENT SETNAME ClientName` command for
     each conn.
     This can also be specified via the 'PODINFO_REDIS_CLIENT_NAME' env
     variable

  -redis-conn-max-idle-time=<value>
     flag used to set the maximum amount of time a connection may be
     idle.Should be less than server's timeout
     Expired connections may be closed lazily before reuseIf d <= 0,
     connections are not closed due to a connection's idle time.
     Default is 30 minutes. -1 disables idle timeout check.
     This can also be specified via the
     'PODINFO_REDIS_CONN_MAX_IDLE_TIME' env variable

  -redis-conn-max-lifetime=<value>
     flag used to set the maximum amount of time a connection may be
     reused.
     Expired connections may be closed lazily before reuse.If <= 0,
     connections are not closed due to a connection's age.
     Default is to not close idle connections.
     This can also be specified via the
     'PODINFO_REDIS_CONN_MAX_LIFETIME' env variable

  -redis-context-timeout-enabled
     flag used to set  context timeout enabled.
     contextTimeoutEnabled controls whether the client respects context
     timeouts and deadlines.
     See https://.uptrace.dev/guide/go--debugging.html#timeouts
     This can also be specified via the
     'PODINFO_REDIS_CONTEXT_TIMEOUT_ENABLED' env variable

  -redis-db=<value>
     flag used to set  database.
     Database to be selected after connecting to the server.
     This can also be specified via the 'PODINFO_REDIS_db' env variable

  -redis-dial-timeout=<0>
     flag used to set timeout for socket reads.
     If reached, commands will failwith a timeout instead of blocking.
     Supported values:
     - `0` - default timeout (3 seconds).
     - `-1` - no timeout (block indefinitely).
     - `-2` - disables SetReadDeadline calls completely.
     Default is 5 seconds.
     This can also be specified via the 'PODINFO_REDIS_DIAL_TIMEOUT' env
     variable

  -redis-max-idle-conns=<value>
     flag used to set the maximum number of idle connections.
     This can also be specified via the 'PODINFO_REDIS_MAX_IDLE_CONNS'
     env variable

  -redis-max-retries=<value>
     flag used to set the maximum number of retries before giving up.
     Default is 3 retries; -1 disables retries.
     This can also be specified via the 'PODINFO_REDIS_MAX_RETRIES' env
     variable

  -redis-max-retry-backoff=<value>
     flag used to set Specifies the maximum backoff between each retry.
     Default is 512 milliseconds; -1 disables backoff.
     This can also be specified via the
     'PODINFO_REDIS_MAX_RETRY_BACKOFF' env variable

  -redis-min-idle-conns=<value>
     flag used to set the minimum number of idle connections.
     it is useful when establishing new connection is slow.
     This can also be specified via the 'PODINFO_REDIS_MIN_IDLE_CONNS'
     env variable

  -redis-min-retry-backoff=<value>
     flag used to set minimum backoff between each retry.
     Default is 8 milliseconds; -1 disables backoff.
     This can also be specified via the
     'PODINFO_REDIS_MIN_RETRY_BACKOFF' env variable

  -redis-password=<value>
     flag used to set  password.
     Optional password. Must match the password specified in
     therequirepass server configuration option (if connecting to a  5.0
     instance, or lower),or the User password when connecting to a  6.0
     instance, or greater,that is using the  ACL system.
     This can also be specified via the 'PODINFO_REDIS_PASSWORD' env
     variable

  -redis-pool-fifo
     flag used to set  pool fifo.
     Type of connection pool.
     true for FIFO pool, false for LIFO pool.
     Note that FIFO has slightly higher overhead compared to LIFO,but it
     helps closing idle connections faster reducing the pool size.
     This can also be specified via the 'PODINFO_REDIS_POOL_FIFO' env
     variable

  -redis-pool-size=<value>
     flag used to set  pool size.
     Maximum number of socket connections.
     Default is 10 connections per every available CPU
     This can also be specified via the 'PODINFO_REDIS_POOL_SIZE' env
     variable

  -redis-pool-timeout=<value>
     flag used to set  pool timeout.
     Amount of time client waits for connection if all connectionsare
     busy before returning an error.
     Default is readTimeout + 1 second.
     This can also be specified via the 'PODINFO_REDIS_POOL_TIMEOUT' env
     variable

  -redis-read-timeout=<value>
     flag used to set  read timeout.
     Timeout for socket reads. If reached, commands will failwith a
     timeout instead of blocking. Supported values:
     - '0' - default timeout (3 seconds).
     - '-1' - no timeout (block indefinitely).
     - '-2' - disables SetReadDeadline calls completely.
     This can also be specified via the 'PODINFO_REDIS_READ_TIMEOUT' env
     variable

  -redis-username=<value>
     flag used to set  username.
     Use the specified username to authenticate the current
     connectionwith one of the connections defined in the ACL list when
     connectingto a  6.0 instance, or greater, that is using the  ACL
     system.
     This can also be specified via the 'PODINFO_REDIS_USERNAME' env
     variable

  -redis-write-timeout=<0>
     flag used to set  write timeout.
     Timeout for socket writes. If reached, commands will failwith a
     timeout instead of blocking.  Supported values:
     - `0` - default timeout (3 seconds).
     - `-1` - no timeout (block indefinitely).
     - `-2` - disables SetWriteDeadline calls completely.
     This can also be specified via the 'PODINFO_REDIS_WRITE_TIMEOUT'
     env variable
```

## Code Statistics

```console
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 Dockerfile              2          657          567           89            1
 Go                    243        17554        13492         2553         1509
 HCL                     1          129           49           80            0
 JavaScript              1           25           23            2            0
 JSON                    5          291          291            0            0
 Shell                   5          205          179           22            4
 Plain Text              2          193            0          193            0
 TOML                    1           57           57            0            0
 YAML                   10          943          864           71            8
-------------------------------------------------------------------------------
 Markdown                3         1395            0         1138          257
 |- BASH                 1           25           22            1            2
 (Total)                           1420           22         1139          259
===============================================================================
 Total                 273        21449        15522         4148         1779
===============================================================================
```

## Usage and Demo

### Docker Compose Environment

- You can use the docker-compose file to bring up the environment or you can
  build it from source.

```bash
docker compose up -d
```

### Build from source

- Build The Binary

```bash
❯ just build
                      _   _            __
  _ __     ___     __| | (_)  _ __    / _|   ___
 | '_ \   / _ \   / _` | | | | '_ \  | |_   / _ \
 | |_) | | (_) | | (_| | | | | | | | |  _| | (_) |
 | .__/   \___/   \__,_| |_| |_| |_| |_|    \___/
 |_|

# Build Command ------------------------------------------

go build \
  -ldflags \
  '

CHANGELOG.md
# Changelog

### [v1.0.1][1] (2023-02-21)

#### Documentation

- **changelog:** v1.0.0 ([eb5e9f3][2])

#### Chore

- package rename ([3a129a3][3])
- **docker-compose:** add `podinfo` service ([32a7e7d][4])

## [v1.0.0][5] (2023-02-15)

### Documentation

- **changelog:** v0.9.0 ([8ead18d][6])

### Build System

- **just/podinfo:** `/cache/` api group recipes ([2108ed4][7])

## [v0.9.0][8] (2023-02-15)

### Bug Fixes

- **api/handlers/cache:** all cache endpoints are working ([63e63c7][9])
- **api/handlers/cache:** enabled `/cache` api group ([718eae8][10])

### Documentation

- **changelog:** v0.9.0 ([9f50cc4][11])

### Refactors

- better logging ([f61c835][12])
- **build/go/version:** remove unnecessary information ([6999c69][13])

### Chore

- **docker-compose:** fix redis listener ([3ad11d9][14])
- **docker-compose:** `redis` service configuration file ([d3178be][15])
- **docker-compose:** `redis` service ([5f696a1][16])

### [v0.8.2][17] (2023-02-15)

#### Documentation

- **changelog:** v0.9.0 ([868a12d][18])

#### CI

- **docker:** checking to see if docker credenials exist ([2e7d3fa][19])
- **docker:** checking to see if docker credenials exist ([06a0f86][20])
- **docker:** trigger build after adding secrets ([e8be68d][21])
- **docker:** fix `buildx` id collision ([6f1f2a8][22])

### [v0.8.1][23] (2023-02-15)

#### Documentation

- **changelog:** v0.8.0 ([9522755][24])

#### CI

- **dockeer:** trigger build on changes to any `*.go` file ([7ec1748][25])

## [v0.8.0][26] (2023-02-15)

### Documentation

- add some instructions and context ([344ff9a][27])
- update `just` target in main `README` file ([68e9779][28])
- add `Roadmap` ([a3b0242][29])
- `swagger` yaml ([efb738c][30])
- **changelog:** v0.7.0 ([13a5032][31])

## [v0.7.0][32] (2023-02-15)

### Features

- **cmd/podinfo/commands/server:** parse redis info with flags ([8e57115][33])
- **api/handlers/cache:** register **DELETE** `/cache/{key}` route in API group
  `Register()` ([a5ab3a8][34])
- **api/handlers/cache:** register **DELETE** `/cache/{key}` route in API group
  `Register()` ([d12f7db][35])
- **api/handlers/cache/delete:** `handler` implementation ([07c94d5][36])

### Documentation

- added code stats ([3e0e1ef][37])
- **api/handlers/cache/delete:** package synopsis ([d0e25cb][38])
- **changelog:** v0.7.0 ([8b5e5d2][39])

### [v0.6.3][40] (2023-02-15)

#### Features

- **api/handlers/cache:** register **GET** `/cache/{key}` route in API group
  `Register()` ([ab70334][41])
- **api/handlers/cache/get:** route `register` implementation ([9935643][42])
- **api/handlers/cache/get:** `handler` implementation ([f154df8][43])

#### Documentation

- **api/handlers/cache/doc:** package synopsis ([9de9d35][44])
- **changelog:** v0.7.0 ([61739b5][45])

### [v0.6.2][46] (2023-02-15)

#### Features

- **api/handlers/cache:** register **PUT** `/cache/{key}` route in API group
  `Register()` ([b7cb2ba][47])
- **api/handlers/cache/put:** route `register` implementation ([6696e26][48])
- **api/handlers/cache/put:** `handler` implementation ([15a177c][49])

#### Documentation

- **api/handlers/cache/put:** package synopsis ([26257c2][50])
- **changelog:** v0.7.0 ([1b6e7f1][51])

### [v0.6.1][52] (2023-02-15)

#### Features

- **api/handlers/cache:** register **POST** `/cache/{key}` route in API group
  `Register()` ([f2f2164][53])
- **api/handlers/cache/post:** route `register` implementation ([fb15f04][54])
- **api/handlers/cache/post:** `handler` implementation ([86f9e9d][55])

#### Documentation

- **api/handlers/cache:** package synopsis ([1e8dc7e][56])
- **api/handlers/cache/post:** package synopsis ([f878c09][57])
- **changelog:** v0.6.0 ([cf372a5][58])

## [v0.6.0][59] (2023-02-15)

### Features

- **api/handlers/cache/shared:** `preflight` request handler ([489db81][60])
- **api/handlers/cache/shared:** `API group` shared constructs ([47b687d][61])
- **cmd/podinfo/commands/server:** flags to read `redis` config ([76a07eb][62])
- **api/core:** add `redis` configuration fields and methods ([cc54fcf][63])
- **internal/cli/value:** handle `int` type cli flags ([d26c406][64])

### Documentation

- **api/handlers/cache/shared:** package synopsis ([2732306][65])
- **changelog:** v0.5.0 ([c313777][66])

### Refactors

- **internal/cli/flagset:** more flexibilty in prettifying generated flag usage
  ([893a1e2][67])
- **internal/cli/flagset:** more flexiblity in self-documenting flags
  ([fc63a46][68])

### Build System

- **docker/podinfo:** minor changes to build script ([9e7f04a][69])

### Chore

- **re-commit:** hook updates ([bff048a][70])

## [v0.5.0][71] (2023-02-15)

### [v0.4.1][72] (2023-02-15)

#### Features

- **api/handlers:** register **GET** `/delay/{seconds}` route in `Initialize()`
  ([5a9dc8e][73])
- **api/handlers/delay:** route `register` implementation ([f39830b][74])
- **api/handlers/delay:** `handler` implementation ([bf469a3][75])

api/core/config.go
package core

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"sync"
	"time"

	logger "github.com/da-moon/podinfo/internal/logger"
	metrics "github.com/da-moon/podinfo/sdk/api/metrics"
	port "github.com/da-moon/podinfo/sdk/api/port"
	mapstructure "github.com/mitchellh/mapstructure"
	stacktrace "github.com/palantir/stacktrace"
)

// config is the configuration for the agent.
//
//go:generate gomodifytags -override -file $GOFILE -struct config -add-tags json,yaml,mapstructure -w -transform snakecase
type Config struct {
	mutex           sync.RWMutex          `json:"lock" yaml:"lock" mapstructure:"lock"`
	log             *logger.WrappedLogger `json:"log" yaml:"log" mapstructure:"log"`
	DevelopmentMode bool                  `mapstructure:"development_mode" json:"development_mode,omitempty" yaml:"development_mode"`
	NodeName        string                `mapstructure:"node_name" json:"node_name,omitempty" yaml:"node_name"`
	APIAddr         string                `mapstructure:"api_addr" json:"api_addr,omitempty" yaml:"api_addr"`
	// ─── METRICS ────────────────────────────────────────────────────────────────────
	MetricsPrefix           string        `mapstructure:"metrics_prefix" json:"metrics_prefix" yaml:"metrics_prefix"`
	StatsiteAddr            string        `mapstructure:"statsite_addr" json:"statsite_addr,omitempty" yaml:"statsite_addr"`
	StatsdAddr              string        `mapstructure:"statsd_addr" json:"statsd_addr,omitempty" yaml:"statsd_addr"`
	PrometheusRetentionTime time.Duration `mapstructure:"prometheus_retention_time" json:"prometheus_retention_time" yaml:"prometheus_retention_time"`
	// ─── REDIS ───────────────────────────────────────────────────────────────────
	// host:port address.
	RedisAddr string
	// ClientName will execute the `CLIENT SETNAME ClientName` command for each conn.
	RedisClientName string
	// Use the specified Username to authenticate the current connection
	// with one of the connections defined in the ACL list when connecting
	// to a Redis 6.0 instance, or greater, that is using the Redis ACL system.
	RedisUsername string
	// Optional password. Must match the password specified in the
	// requirepass server configuration option (if connecting to a Redis 5.0 instance, or lower),
	// or the User Password when connecting to a Redis 6.0 instance, or greater,
	// that is using the Redis ACL system.
	RedisPassword string
	// Database to be selected after connecting to the server.
	RedisDB int
	// Maximum number of retries before giving up.
	// -1 disables retries.
	RedisMaxRetries int
	// Minimum backoff between each retry.
	// -1 disables backoff.
	RedisMinRetryBackoff time.Duration
	// Maximum backoff between each retry.
	// -1 disables backoff.
	RedisMaxRetryBackoff time.Duration
	// Dial timeout for establishing new connections.
	RedisDialTimeout time.Duration
	// Timeout for socket reads. If reached, commands will fail
	// with a timeout instead of blocking. Supported values:
	//   - `0` - default timeout (3 seconds).
	//   - `-1` - no timeout (block indefinitely).
	//   - `-2` - disables SetReadDeadline calls completely.
	RedisReadTimeout time.Duration
	// Timeout for socket writes. If reached, commands will fail
	// with a timeout instead of blocking.  Supported values:
	//   - `0` - default timeout (3 seconds).
	//   - `-1` - no timeout (block indefinitely).
	//   - `-2` - disables SetWriteDeadline calls completely.
	RedisWriteTimeout time.Duration
	// ContextTimeoutEnabled controls whether the client respects context timeouts and deadlines.
	// See https://redis.uptrace.dev/guide/go-redis-debugging.html#timeouts
	RedisContextTimeoutEnabled bool
	// Type of connection pool.
	// true for FIFO pool, false for LIFO pool.
	// Note that FIFO has slightly higher overhead compared to LIFO,
	// but it helps closing idle connections faster reducing the pool size.
	RedisPoolFIFO bool
	// Maximum number of socket connections.
	RedisPoolSize int
	// Amount of time client waits for connection if all connections
	// are busy before returning an error.
	RedisPoolTimeout time.Duration
	// Minimum number of idle connections which is useful when establishing
	// new connection is slow.
	RedisMinIdleConns int
	// Maximum number of idle connections.
	RedisMaxIdleConns int
	// ConnMaxIdleTime is the maximum amount of time a connection may be idle.
	// Should be less than server's timeout.
	RedisConnMaxIdleTime time.Duration
	// ConnMaxLifetime is the maximum amount of time a connection may be reused.
	RedisConnMaxLifetime time.Duration
}

// Mergeconfig takes in two config objects
// and merges them into one, giving precedence to the second config.
// nolint:gocognit,gocyclo // this function is well tested and won't be used often
func Mergeconfig(a, b *Config) *Config { // revive:disable:unexported-return
	result := *a
	result.DevelopmentMode = b.DevelopmentMode
	if b.NodeName != "" {
		result.NodeName = b.NodeName
	}
	if b.APIAddr != "" {
		result.APIAddr = b.APIAddr
	}
	// ─── TELEMETRY ───────────────────────────────────────────────────────────────
	if b.MetricsPrefix != "" {
		result.MetricsPrefix = b.MetricsPrefix
	}
	if b.StatsiteAddr != "" {
		result.StatsiteAddr = b.StatsiteAddr
	}
	if b.StatsdAddr != "" {
		result.StatsdAddr = b.StatsdAddr
	}
	if b.PrometheusRetentionTime.Nanoseconds() >= 1 {
		result.PrometheusRetentionTime = b.PrometheusRetentionTime
	}
	// ─── REDIS
	// ───────────────────────────────────────────────────────────────────
	if b.RedisAddr != "" {
		result.RedisAddr = b.RedisAddr
	}
	if b.RedisClientName != "" {
		result.RedisClientName = b.RedisClientName
	}
	if b.RedisUsername != "" {
		result.RedisUsername = b.RedisUsername
	}
	if b.RedisPassword != "" {
		result.RedisPassword = b.RedisPassword
	}
	result.RedisDB = b.RedisDB
	if b.RedisMaxRetries >= 0 {
		result.RedisMaxRetries = b.RedisMaxRetries
	}
	if b.RedisMinRetryBackoff.Nanoseconds() >= -1 {
		result.RedisMinRetryBackoff = b.RedisMinRetryBackoff
	}
	if b.RedisMaxRetryBackoff.Nanoseconds() >= -1 {
		result.RedisMaxRetryBackoff = b.RedisMaxRetryBackoff
	}
	if b.RedisDialTimeout.Nanoseconds() >= 0 {
		result.RedisDialTimeout = b.RedisDialTimeout
	}
	if b.RedisReadTimeout.Nanoseconds() >= -2 {
		result.RedisReadTimeout = b.RedisReadTimeout
	}
	if b.RedisWriteTimeout.Nanoseconds() >= -2 {
		result.RedisWriteTimeout = b.RedisWriteTimeout
	}
	result.RedisContextTimeoutEnabled = b.RedisContextTimeoutEnabled
	result.RedisPoolFIFO = b.RedisPoolFIFO
	if b.RedisPoolSize >= 0 {
		result.RedisPoolSize = b.RedisPoolSize
	}
	if b.RedisPoolTimeout >= 0 {
		result.RedisPoolTimeout = b.RedisPoolTimeout
	}
	if b.RedisMinIdleConns != 0 {
		result.RedisMinIdleConns = b.RedisMinIdleConns
	}
	if b.RedisMaxIdleConns >= 0 {
		result.RedisMaxIdleConns = b.RedisMaxIdleConns
	}
	if b.RedisConnMaxIdleTime.Nanoseconds() >= 0 {
		result.RedisConnMaxIdleTime = b.RedisConnMaxIdleTime
	}
	result.RedisConnMaxLifetime = b.RedisConnMaxLifetime
	return &result
	// revive:enable:unexported-return
}

//
// ──────────────────────────────────────────────────────────────────── I ──────────
//   :::::: D E F A U L T   V A L U E S : :  :   :    :     :        :          :
// ──────────────────────────────────────────────────────────────────────────────
//

// DefaultConfig returns a new config struct
// nolint:gocognit,gocyclo // this function is well tested and won't be used often
func DefaultConfig(log *logger.WrappedLogger) (*Config, error) { // revive:disable:unexported-return
	if log == nil {
		err := stacktrace.NewError("no logger was provided")
		return nil, err
	}
	// ─── DEVELOPMENT MODE DEFAULT ───────────────────────────────────────────────────
	developmentMode := DefaultDevelopmentMode()
	// ─── NODE NAME DEFAULT ──────────────────────────────────────────────────────────
	nodeName, err := DefaultNodeName()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	// ─── API ADDRESS DEFAULT ─────────────────────────────────────────────────────────
	apiAddr, err := DefaultAPIAddr()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	// ─── REDIS DEFAULTS ──────────────────────────────────────────────────────────
	redisAddr, err := DefaultRedisAddr()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisClientName, err := DefaultRedisClientName()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	// No need to throw an error
	redisUsername, _ := DefaultRedisUsername()
	redisPassword, _ := DefaultRedisPassword()
	redisDB, err := DefaultRedisDB()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisMaxRetries, err := DefaultRedisMaxRetries()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisMinRetryBackoff, err := DefaultRedisMinRetryBackoff()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisMaxRetryBackoff, err := DefaultRedisMaxRetryBackoff()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisDialTimeout, err := DefaultRedisDialTimeout()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisReadTimeout, err := DefaultRedisReadTimeout()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisWriteTimeout, err := DefaultRedisWriteTimeout()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisContextTimeoutEnabled, err := DefaultRedisContextTimeoutEnabled()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisPoolFIFO, err := DefaultRedisPoolFIFO()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisPoolSize, err := DefaultRedisPoolSize()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisPoolTimeout, err := DefaultRedisPoolTimeout()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisMinIdleConns, err := DefaultRedisMinIdleConns()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisMaxIdleConns, err := DefaultRedisMaxIdleConns()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisConnMaxIdleTime, err := DefaultRedisConnMaxIdleTime()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	redisConnMaxLifetime, err := DefaultRedisConnMaxLifetime()
	if err != nil {
		err = stacktrace.Propagate(err, "cannot prepare default api config struct")
		return nil, err
	}
	// ─── RESULT ─────────────────────────────────────────────────────────────────────
	// ────────────────────────────────────────────────────────────────────────────────
	// ────────────────────────────────────────────────────────────────────────────────
	result := &Config{
		log:             log,
		DevelopmentMode: developmentMode,
		NodeName:        nodeName,
		APIAddr:         apiAddr,
		// ─── REDIS ────────────────────────────────────────────────────
		RedisAddr:                  redisAddr,
		RedisClientName:            redisClientName,
		RedisUsername:              redisUsername,
		RedisPassword:              redisPassword,
		RedisDB:                    redisDB,
		RedisMaxRetries:            redisMaxRetries,
		RedisMinRetryBackoff:       redisMinRetryBackoff,
		RedisMaxRetryBackoff:       redisMaxRetryBackoff,
		RedisDialTimeout:           redisDialTimeout,
		RedisReadTimeout:           redisReadTimeout,
		RedisWriteTimeout:          redisWriteTimeout,
		RedisContextTimeoutEnabled: redisContextTimeoutEnabled,
		RedisPoolFIFO:              redisPoolFIFO,
		RedisPoolSize:              redisPoolSize,
		RedisPoolTimeout:           redisPoolTimeout,
		RedisMinIdleConns:          redisMinIdleConns,
		RedisMaxIdleConns:          redisMaxIdleConns,
		RedisConnMaxIdleTime:       redisConnMaxIdleTime,
		RedisConnMaxLifetime:       redisConnMaxLifetime,
		// ─── METRICS ─────────────────────────────────────────────────────
		MetricsPrefix:           metrics.DefaultMetricsPrefix(),
		StatsiteAddr:            metrics.DefaultStatsiteAddr(),
		StatsdAddr:              metrics.DefaultStatsdAddr(),
		PrometheusRetentionTime: metrics.DefaultPrometheusRetentionTime(),
	}
	return result, nil
	// revive:enable:unexported-return
}

// DefaultDevelopmentMode returns the default value for the development mode
func DefaultDevelopmentMode() bool {
	var result bool
	podinfoDevelString := os.Getenv("PODINFO_DEVEL")
	if podinfoDevelString != "" {
		var err error
		result, err = strconv.ParseBool(podinfoDevelString)
		if err != nil {
			result = false
		}
	}
	return result
}

// DefaultNodeName returns the default value for the node name
func DefaultNodeName() (string, error) {
	var err error
	result := os.Getenv("PODINFO_NODE_NAME")
	if result == "" {
		result, err = os.Hostname()
		if err != nil {
			err = stacktrace.Propagate(err, "cannot get default node name")
			return "", err
		}
	}
	return result, nil
}

// DefaultAPIAddr returns the default value for the api address
// TODO: add validation to ensure address is valid when the user provided the address
func DefaultAPIAddr() (string, error) {
	apiAddr := os.Getenv("PODINFO_API_ADDR")
	if apiAddr == "" {
		tcpAddr, err := port.New().TCP()
		if err != nil {
			err = stacktrace.Propagate(err, "cannot find a random port to bind to")
			return "", err
		}
		apiAddr = fmt.Sprintf("0.0.0.0:%s", tcpAddr.Port)
	}
	return apiAddr, nil
}

//
// ──────────────────────────────────────────────────── I ──────────
//   :::::: G E T T E R : :  :   :    :     :        :          :
// ──────────────────────────────────────────────────────────────
//

// Decodeconfig takes an io.reader and decodes
// underlying byte stream into a config struct
// nolint:gocognit // This is a well tested function
func Decodeconfig(r io.Reader) (*Config, error) { // revive:disable:unexported-return
	var raw interface{}
	dec := json.NewDecoder(r)
	if err := dec.Decode(&raw); err != nil {
		return nil, err
	}
	var md mapstructure.Metadata
	var result Config
	msdec, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
		Metadata:    &md,
		Result:      &result,
		ErrorUnused: true,
	})
	if err != nil {
		return nil, err
	}
	if err := msdec.Decode(raw); err != nil {
		return nil, err
	}
	return &result, nil
	// revive:enable:unexported-return
}

// Log returns the underlying log writer
func (c *Config) Log() *logger.WrappedLogger {
	c.mutex.RLock()
	return c.log
}

// ──────────────────────────────────────────────────── I ──────────
//
//	:::::: S E T T E R : :  :   :    :     :        :          :
//
// ──────────────────────────────────────────────────────────────
func (c *Config) SetDevelopmentMode() {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	c.DevelopmentMode = true
}

func (c *Config) SetNodeName(value string) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	if value != "" {
		c.NodeName = value
	}
}
func (c *Config) SetAPIAddr(value string) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	if value != "" {
		c.APIAddr = value
	}
}
func (c *Config) SetMetricsPrefix(value string) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	if value != "" {
		c.MetricsPrefix = value
	}
}
func (c *Config) SetStatsiteAddr(value string) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	if value != "" {
		c.StatsiteAddr = value
	}
}
func (c *Config) SetStatsdAddr(value string) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	if value != "" {
		c.StatsdAddr = value
	}
}
func (c *Config) SetPrometheusRetentionTime(value time.Duration) {
	c.mutex.Lock()
	defer c.mutex.Unlock()
	if value > 0 {
		c.PrometheusRetentionTime = value
	}
}

//
// ────────────────────────────────────────────────────────────── I ──────────
//   :::::: F I L E   L O A D E R : :  :   :    :     :        :          :
// ────────────────────────────────────────────────────────────────────────
//

// ReadconfigPaths takes in an array of file (json)
// paths and returns a config struct.
// TODO: read configs in parallel with help of a waitgroup as calling defer in loop might lead to bugs
// nolint:gocognit,gocyclo // I can't think of any ways to lower complexity
func ReadconfigPaths(paths []string) (*Config, error) { // revive:disable:unexported-return,defer
	result := &Config{
		log: logger.DefaultWrappedLogger(string(logger.ErrorLevel)),
	}
	for _, p := range paths {
		path := filepath.Clean(p)
		f, err := os.Open(path)
		if err != nil {
			return nil, stacktrace.Propagate(err, "error reading '%s'", path)
		}
		// note(damoon) this may cause the application to open many
		// file descriptors which might cause the kernel to panic
		if f != nil {
			defer func() {
				err = f.Close()
				if err != nil {
					result.log.Error("cannot close file:%v", err)
				}
			}()
		}
		fi, err := f.Stat()
		if err != nil {
			return nil, stacktrace.Propagate(err, "error reading '%s'", path)
		}
		if !fi.IsDir() {
			dec := new(Config) // nolint:staticcheck // SA4006 dec is used right away
			dec, err = Decodeconfig(f)
			if err != nil {
				return nil, stacktrace.Propagate(err, "error decoding '%s'", path)
			}
			result = Mergeconfig(result, dec)
			continue
		}
		contents, err := f.Readdir(-1)
		if err != nil {
			return nil, stacktrace.Propagate(err, "error reading '%s'", path)
		}
		sort.Sort(dirEnts(contents))
		for _, fi := range contents {
			if fi.IsDir() {
				continue
			}
			if !strings.HasSuffix(fi.Name(), ".json") {
				continue
			}
			subpath := filepath.Clean(
				filepath.Join(path, fi.Name()),
			)
			ff, err := os.Open(subpath)
			if err != nil {
				return nil, stacktrace.Propagate(err, "error reading '%s'", subpath)
			}
			if ff != nil {
				defer func() {
					err = ff.Close()
					if err != nil {
						result.log.Error("cannot close file:%v", err)
					}
				}()
			}
			config, err := Decodeconfig(ff)
			if err != nil {
				return nil, stacktrace.Propagate(err, "error decoding '%s'", subpath)
			}
			result = Mergeconfig(result, config)
		}
	}
	return result, nil
	// revive:enable:unexported-return,defer
}

// ─── SORT INTERFACE ─────────────────────────────────────────────────────────────
type dirEnts []os.FileInfo

// Len implements sort.Interface.
func (d dirEnts) Len() int { return len(d) }

// Less implements sort.Interface
func (d dirEnts) Less(i, j int) bool { return d[i].Name() < d[j].Name() }

// Swap implement sort.Interface
func (d dirEnts) Swap(i, j int) { d[i], d[j] = d[j], d[i] }

// ────────────────────────────────────────────────────────────────────────────────
